apps/src/JSDebuggerUI.js

```diff
@@ -0,0 +1,502 @@
+  }
+};
+
+/**
+ * Get the step delay in milliseconds from the speed slider in the debugger UI.
+ * If no speed slider is present, returns undefined.
+ * @return {number|undefined}
+ */
+JSDebuggerUI.prototype.getStepDelay = function () {
+  if (this.speedSlider_) {
+    return JSDebuggerUI.stepDelayFromSliderPercent(this.speedSlider_.getValue());
+  }
+  return undefined;
+};
+
+/**
+ * Set the speed slider position.
+ * @param {!number} percent - range 0.0-1.0
+ */
+JSDebuggerUI.prototype.setStepSpeedPercent = function (percent) {
+  if (this.speedSlider_) {
+    this.speedSlider_.setValue(percent);
+  }
+};
+
+/**
+ * Exponential conversion from slider position as a percentile to a step delay
+ * in milliseconds.
+ * @param {!number} stepSpeedPercentage range 0.0-1.0
+ * @returns {number} step delay in milliseconds
+ */
+JSDebuggerUI.stepDelayFromSliderPercent = function (stepSpeedPercentage) {
+  return 300 * Math.pow(1 - stepSpeedPercentage, 2);
+};
+
+/**
+ * Given some object or message, attempt to log it both to the browser console
+ * and to the user-facing debug console.
+ * @param {*} output
+ */
+JSDebuggerUI.prototype.log = function (output) {
+  // first pass through to the real browser console log if available:
+  if (console && console.log) {
+    console.log(output);
+  }
+
+  // then put it in the debug console visible to the user:
+  var debugOutputDiv = this.getElement_('#debug-output');
+  if (debugOutputDiv) {
+    if (debugOutputDiv.textContent.length > 0) {
+      debugOutputDiv.textContent += '\n';
+    }
+    debugOutputDiv.textContent += stringifyNonStrings(output);
+
+    debugOutputDiv.scrollTop = debugOutputDiv.scrollHeight;
+  }
+};
+
+/**
+ * @param {*} object
+ * @returns {string}
+ */
+function stringifyNonStrings(object) {
+  if (typeof object === 'string' || object instanceof String) {
+    return object;
+  } else {
+    return JSON.stringify(object);
+  }
+}
+
+/**
+ * Handler for key events in the debug console input box.
+ * @param {KeyboardEvent} e
+ */
+JSDebuggerUI.prototype.onDebugInputKeyDown = function (e) {
+  var input = e.target.textContent;
+  if (e.keyCode === KeyCodes.ENTER) {
+    e.preventDefault();
+    pushDebugConsoleHistory(input);
+    e.target.textContent = '';
+    this.log('> ' + input);
+    var jsInterpreter = this.getJSInterpreter_();
+    if (jsInterpreter) {
+      try {
+        var result = jsInterpreter.evalInCurrentScope(input);
+        this.log('< ' + String(result));
+      } catch (err) {
+        this.log('< ' + String(err));
+      }
+    } else {
+      this.log('< (not running)');
+    }
+  }
+  if (e.keyCode === KeyCodes.UP) {
+    updateDebugConsoleHistory(input);
+    e.target.textContent = moveUpDebugConsoleHistory(input);
+  }
+  if (e.keyCode === KeyCodes.DOWN) {
+    updateDebugConsoleHistory(input);
+    e.target.textContent = moveDownDebugConsoleHistory(input);
+  }
+};
+
+//Debug console history
+var consoleHistory = {
+  history: [],
+  currentIndex: 0
+};
```

Comment:
I'm curious if you considered making this a member of JSDebuggerUI? Ironically, not making it a member (as you've done) actually protects it from outside access